vopono_core/src/network/piapf.rs

```diff
@@ -0,0 +1,193 @@
+            log::error!("Signature for port forward from PIA API not OK");
+            anyhow::bail!("Signature for port forward from PIA API not OK");
+        }
+        
+        let signature = parsed["signature"].as_str().unwrap().to_string();
+        let payload = parsed["payload"].as_str().unwrap().to_string();
+        let decoded = BASE64_STANDARD.decode(&payload)?;
+        let parsed = json::parse(String::from_utf8_lossy(&decoded).as_ref())?;
+        let port = parsed["port"].as_u16().unwrap();
+
+        let params = ThreadParams {
+            netns_name: ns.name.clone(),
+            hostname: vpn_hostname,
+            gateway: vpn_gateway,
+            signature: signature,
+            payload: payload,
+            port: port,
+        };
+        Self::refresh_port(&params)?;
+        let (send, recv) = mpsc::channel::<bool>();
+        let handle = std::thread::spawn(move || Self::thread_loop(params, recv));
+
+        log::info!("PIA forwarded local port: {port}");
+        Ok(Self {
+            port: port,
+            loop_thread_handle: Some(handle),
+            send_channel: send,
+        })
+    }
+
+    fn refresh_port(params: &ThreadParams) -> anyhow::Result<u16> {
+    
+        let bind_response = NetworkNamespace::exec_with_output(&params.netns_name, &["curl", 
+            "-Gs", "-m", "5",
+            "--connect-to", &format!("{}::{}:", params.hostname, params.gateway).to_string(),
+            "--cacert", "/home/benland100/vopono/vopono_core/src/config/providers/pia/ca.rsa.4096.crt", //FIXME: how to get this path?
+            "--data-urlencode", &format!("payload={}", params.payload).to_string(),
+            "--data-urlencode", &format!("signature={}", params.signature).to_string(),
+            &format!("https://{}:19999/bindPort", params.hostname).to_string() ], )?;
+        if !bind_response.status.success() {
+            log::error!("Could not bind port forward from PIA API");
+            anyhow::bail!("Could not bind port forward from PIA API")
+        }
+        
+        let parsed = json::parse(String::from_utf8_lossy(&bind_response.stdout).as_ref())?;
+        
+        if parsed["status"] != "OK" {
+            log::error!("Bind for port forward from PIA API not OK");
+            anyhow::bail!("Bind for port forward from PIA API not OK");
+        }
+        
+        //FIXME: its very useful to have a configurable callback script to receive the port number
+        let refresh_response = NetworkNamespace::exec_with_output(&params.netns_name, &["/home/benland100/vopono/test_callback.sh", &params.port.to_string()], )?;
+        if !refresh_response.status.success() {
+            log::info!("Callback script was unsuccessful!");
+        }
+        
+        log::info!("Successfully updated claim to port {}", params.port);
+
+        Ok(params.port)
+    }
+
+    // Spawn thread to repeat above every 15 minutes
+    fn thread_loop(params: ThreadParams, recv: Receiver<bool>) {
```

Comment:
We might be able to put this and the channel logic in a trait to share it a bit better with natpmpc but this isn't critical (and could be tricky with the parts shared across threads).